### PR TITLE
[DRILL-8457] Allow configuring csv parser in http storage plugin configuration

### PR DESCRIPTION
**Pull request by ztomanek-dw**
# [DRILL-8457](https://issues.apache.org/jira/browse/DRILL-8457): Allow configuring csv parser in http storage plugin configuration

## Description

HttpApiConfiguration was extended with `csvOptions` field which allows setting a following properties:

```json
{
  "csvOptions": {
    "delimiter": ",",
    "quote": "\"",
    "quoteEscape": "\"",
    "lineSeparator": "\n",
    "headerExtractionEnabled": null,
    "numberOfRowsToSkip": 0,
    "numberOfRecordsToRead": -1,
    "lineSeparatorDetectionEnabled": true,
    "maxColumns": 512,
    "maxCharsPerColumn": 4096,
    "skipEmptyLines": true,
    "ignoreLeadingWhitespaces": true,
    "ignoreTrailingWhitespaces": true,
    "nullValue": null
  }
}
```

this provides greater csv parsing flexibility since user can set different delimiters, number of columns or max column size. 

Also backward compatibility is ensured and parser works same as before if `csvOptions` is null.

## Documentation

Add a following paragraph into https://drill.apache.org/docs/http-storage-plugin/#configuring-the-api-connections

##### CSV parser options

CSV parser of HTTP Storage plugin can be configured using `csvOptions`.

```json
{
  "csvOptions": {
    "delimiter": ",",
    "quote": "\"",
    "quoteEscape": "\"",
    "lineSeparator": "\n",
    "headerExtractionEnabled": null,
    "numberOfRowsToSkip": 0,
    "numberOfRecordsToRead": -1,
    "lineSeparatorDetectionEnabled": true,
    "maxColumns": 512,
    "maxCharsPerColumn": 4096,
    "skipEmptyLines": true,
    "ignoreLeadingWhitespaces": true,
    "ignoreTrailingWhitespaces": true,
    "nullValue": null
  }
}
```

E.g. to parse `.tsv` files you can use a following config:

```json
{
  "csvOptions": {
    "delimiter": "\t"
  }
}
```

## Testing

Create a following storage plugin with name `github`

```json
{
  "type": "http",
  "connections": {
    "test-data": {
      "url": "https://raw.githubusercontent.com/semantic-web-company/wic-tsv/master/data/de/Test/test_examples.txt",
      "requireTail": false,
      "method": "GET",
      "authType": "none",
      "inputType": "csv",
      "xmlDataLevel": 1,
      "postParameterLocation": "QUERY_STRING",
      "csvOptions": {
        "delimiter": "\t",
        "quote": "\"",
        "quoteEscape": "\"",
        "lineSeparator": "\n",
        "numberOfRecordsToRead": -1,
        "lineSeparatorDetectionEnabled": true,
        "maxColumns": 512,
        "maxCharsPerColumn": 4096,
        "skipEmptyLines": true,
        "ignoreLeadingWhitespaces": true,
        "ignoreTrailingWhitespaces": true
      },
      "verifySSLCert": true
    }
  },
  "timeout": 5,
  "retryDelay": 1000,
  "proxyType": "direct",
  "authMode": "SHARED_USER",
  "enabled": true
}
```

Then query tsv file with 

```sql
SELECT * from github.`test-data`
```

You should see a result set containing three columns

### CODE REVIEW

**Comment by ztomanek-dw** [comment]
@cgivre 
Thanks for your feedback!

According to your comments:
 - written unit tests for `HttpCSVOptions`
 - written unit tests for `HttpApiConfig`, by the way fixing small bug on `HttpMethod` validation
 - added tsv parsing test to `TestHttpPlugin` 
 - documented `csvOptions` configuration in `CSV_Options.md`

Let me know if you see anything else to cover :)

File: contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/HttpCSVOptions.java [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+    HttpCSVOptions that = (HttpCSVOptions) o;
+    return quote == that.quote && quoteEscape == that.quoteEscape && numberOfRowsToSkip == that.numberOfRowsToSkip && numberOfRecordsToRead == that.numberOfRecordsToRead && lineSeparatorDetectionEnabled == that.lineSeparatorDetectionEnabled && maxColumns == that.maxColumns && maxCharsPerColumn == that.maxCharsPerColumn && skipEmptyLines == that.skipEmptyLines && ignoreLeadingWhitespaces == that.ignoreLeadingWhitespaces && ignoreTrailingWhitespaces == that.ignoreTrailingWhitespaces && delimiter.equals(that.delimiter) && lineSeparator.equals(that.lineSeparator) && Objects.equals(headerExtractionEnabled, that.headerExtractionEnabled) && nullValue.equals(that.nullValue);
```
Comment (cgivre) [inline contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/HttpCSVOptions.java:162]: Nit:  Please break this up into new lines.

File: contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/HttpCSVOptions.java [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+  public int hashCode() {
+    return Objects.hash(delimiter, quote, quoteEscape, lineSeparator, headerExtractionEnabled,
+        numberOfRowsToSkip, numberOfRecordsToRead, lineSeparatorDetectionEnabled, maxColumns,
+        maxCharsPerColumn, skipEmptyLines, ignoreLeadingWhitespaces, ignoreTrailingWhitespaces,
+        nullValue);
+  }
+
+  @Override
+  public String toString() {
+    return "HttpCSVOptions{" + "delimiter='" + delimiter + '\'' + ", quote=" + quote + ", " +
```
Comment (cgivre) [inline contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/HttpCSVOptions.java:175]: Nit:  Please use the `PlanStringBuilder` for the `toString()` method.

**Review by cgivre — CHANGES_REQUESTED** [review]
@ztomanek-dw 
Thank you for the revisions.  I have two minor nits and other than that, LGTM. 
Just as an FYI we'll take care of squashing all commits when we merge so you don't have to worry about that.  Just please make sure it is rebased on the current master, which I think it is.

**Comment by ztomanek-dw** [comment]
@cgivre 
Thanks for the clarification, I was not sure if I could push multiple commits per one jira issue. 
I've applied your suggestions and made sure it's rebased to current master

**Pull request merged**